### PR TITLE
feat(query-db-collection): add manual write methods for direct state updates

### LINKED ISSUE #294: Add Manual Sync Updates API to @tanstack/query-db-collection
## Background

The `@tanstack/query-db-collection` package integrates TanStack Query with TanStack DB collections, providing automatic synchronization between query results and collection state. Currently, it works by:

1. Using TanStack Query to fetch data via `queryFn`
2. Diffing the fetched results against current collection state using shallow equality checks
3. Writing insert/update/delete operations to sync the changes
4. Automatically refetching after mutations (unless `{ refetch: false }` is returned)

This approach works well for many use cases but has limitations in two key scenarios:

1. **Performance/Efficiency**: After mutations, the default behavior refetches all data, which is inefficient for large datasets or expensive queries
2. **Out-of-band updates**: No way to manually sync data received through other channels (WebSockets, SSE, server-sent events in mutation responses)

## Problem Statement

The current implementation in `/packages/query-db-collection/src/query.ts` only updates the collection through:
- Initial query load
- `collection.utils.refetch()` calls
- Automatic refetch after mutations

There's no API to manually write specific changes to the collection without triggering a full refetch. This forces developers to choose between:
- Accepting the performance cost of full refetches
- Implementing complex workarounds to merge out-of-band data

## Proposed Solution

Add manual sync update methods to the `QueryCollectionUtils` interface that allow direct manipulation of the collection's sync state:

```typescript
export interface QueryCollectionUtils extends UtilsRecord {
  refetch: RefetchFn
  
  // New manual sync methods with overloads
  syncInsert: {
    (item: TItem): void
    (items: TItem[]): void
  }
  syncUpdate: {
    (item: TItem): void
    (items: TItem[]): void
  }
  syncDelete: {
    (key: TKey): void
    (keys: TKey[]): void
  }
  syncUpsert: {
    (item: TItem): void
    (items: TItem[]): void
  }
  
  // Batch sync for multiple operations
  syncBatch: (operations: SyncOperation[]) => void
}

type SyncOperation<TItem = any, TKey = string | number> = 
  | { type: 'insert', items: TItem[] }
  | { type: 'update', items: TItem[] }
  | { type: 'delete', keys: TKey[] }
  | { type: 'upsert', items: TItem[] }
```

### 1. Access to Sync Internals

The sync utilities need access to the collection's internal sync mechanism. The current sync implementation uses:
```typescript
const { begin, write, commit, markReady, collection } = params
```

We'll need to expose these through a closure or store a reference during sync initialization.

### 2. Collection State Requirement

Manual sync operations should only be allowed when the collection is in the `ready` state. This ensures:
- The collection has baseline data to work with
- The sync mechanism is properly initialized
- We avoid race conditions between manual syncs and the initial data load

```typescript
syncInsert: (items: TItem | TItem[]) => {
  if (!syncAPI || collection.status !== 'ready') {
    throw new Error(
      '[QueryCollection] Manual sync operations require the collection to be ready. ' +
      'Call `await collection.preload()` before using manual sync methods.'
    )
  }
  
  // Proceed with sync...
}
```

### 3. Runtime Data Integrity Checks

Manual sync operations must validate data integrity to prevent corruption. These checks replace the natural controls that sync engines provide:

```typescript
syncInsert: (items: TItem | TItem[]) => {
  const itemsArray = Array.isArray(items) ? items : [items]
  
  // Check for duplicate keys
  const newKeys = new Set<TKey>()
  
  for (const item of itemsArray) {
    const key = getKey(item)
    
    // Validate key is valid
    if (key === undefined || key === null || key === '') {
      throw new Error(
        `[QueryCollection] Invalid key returned by getKey: ${String(key)}`
      )
    }
    
    // Check for duplicates within the batch
    if (newKeys.has(key)) {
      throw new Error(
        `[QueryCollection] Duplicate key in insert batch: ${String(key)}`
      )
    }
    newKeys.add(key)
    
    // Check for existing keys in collection
    if (collection.has(key)) {
      throw new Error(
        `[QueryCollection] Cannot insert item with existing key: ${String(key)}. ` +
        'Use syncUpdate or syncUpsert for existing items.'
      )
    }
  }
  
  // Proceed with validated insert...
}

syncUpdate: (items: TItem | TItem[]) => {
  const itemsArray = Array.isArray(items) ? items : [items]
  
  // Verify all items exist
  for (const item of itemsArray) {
    const key = getKey(item)
    
    // Validate key is valid
    if (key === undefined || key === null || key === '') {
      throw new Error(
        `[QueryCollection] Invalid key returned by getKey: ${String(key)}`
      )
    }
    
    if (!collection.has(key)) {
      throw new Error(
        `[QueryCollection] Cannot update non-existent item with key: ${String(key)}. ` +
        'Use syncInsert or syncUpsert for new items.'
      )
    }
  }
  
  // Proceed with validated update...
}

syncDelete: (keys: TKey | TKey[]) => {
  const keysArray = Array.isArray(keys) ? keys : [keys]
  
  // Verify all keys exist
  for (const key of keysArray) {
    if (key === undefined || key === null || key === '') {
      throw new Error(
        `[QueryCollection] Invalid key: ${String(key)}`
      )
    }
    
    if (!collection.has(key)) {
      throw new Error(
        `[QueryCollection] Cannot delete non-existent item with key: ${String(key)}`
      )
    }
  }
  
  // Proceed with validated delete...
}

syncUpsert: (items: TItem | TItem[]) => {
  // No existence checks needed - upsert handles both cases
  // But still validate for duplicates within the batch
  const itemsArray = Array.isArray(items) ? items : [items]
  const keys = new Set<TKey>()
  
  for (const item of itemsArray) {
    const key = getKey(item)
    
    // Validate key is valid
    if (key === undefined || key === null || key === '') {
      throw new Error(
        `[QueryCollection] Invalid key returned by getKey: ${String(key)}`
      )
    }
    
    if (keys.has(key)) {
      throw new Error(
        `[QueryCollection] Duplicate key in upsert batch: ${String(key)}`
      )
    }
    keys.add(key)
  }
  
  // Proceed with upsert...
}

syncBatch: (operations: SyncOperation[]) => {
  // Validate batch operations don't conflict
  const insertKeys = new Set<TKey>()
  const updateKeys = new Set<TKey>()
  const deleteKeys = new Set<TKey>()
  
  // First pass: collect all operations
  for (const op of operations) {
    if (op.type === 'insert') {
      for (const item of op.items) {
        const key = getKey(item)
        insertKeys.add(key)
      }
    } else if (op.type === 'update') {
      for (const item of op.items) {
        const key = getKey(item)
        updateKeys.add(key)
      }
    } else if (op.type === 'delete') {
      for (const key of op.keys) {
        deleteKeys.add(key)
      }
    }
  }
  
  // Check for conflicts between operations
  for (const key of insertKeys) {
    if (updateKeys.has(key) || deleteKeys.has(key)) {
      throw new Error(
        `[QueryCollection] Conflicting operations for key: ${String(key)}`
      )
    }
  }
  
  for (const key of updateKeys) {
    if (deleteKeys.has(key)) {
      throw new Error(
        `[QueryCollection] Cannot update and delete same key: ${String(key)}`
      )
    }
  }
  
  // Validate each operation type
  // ... (apply individual operation validations)
}
```

### 4. Query Cache Consistency

Manual updates will update the TanStack Query cache by default to maintain consistency (similar to how normal syncing works):

```typescript
syncUpdate: (items: TItem | TItem[]) => {
  const itemsArray = Array.isArray(items) ? items : [items]
  
  // Validate first (as shown above)
  // ...
  
  // Update collection via sync mechanism
  begin()
  itemsArray.forEach(item => {
    write({ type: 'update', value: item })
  })
  commit()
  
  // Update query cache by default
  queryClient.setQueryData(queryKey, (oldData) => {
    if (!oldData) return oldData
    
    // Create a map for efficient updates
    const dataMap = new Map(oldData.map(item => [getKey(item), item]))
    
    // Apply updates
    itemsArray.forEach(item => {
      dataMap.set(getKey(item), item)
    })
    
    return Array.from(dataMap.values())
  })
}
```

### 5. Schema Validation

While synced data isn't normally validated, manual sync operations will validate data if a schema is provided. This helps developers catch errors early since manual sync operations are more error-prone:

```typescript
// Validate if schema is provided
if (schema) {
  itemsArray.forEach(item => {
    const result = schema.parse(item)
    if (!result.success) {
      throw new Error(`[QueryCollection] Invalid item: ${result.error}`)
    }
  })
}
```

### 6. Error Handling

Errors in manual sync operations will throw, allowing developers to handle them appropriately:
- Collection not ready
- Duplicate keys (within batch or existing)
- Non-existent keys for update/delete
- Invalid keys returned by getKey
- Schema validation errors
- Conflicting operations in batch

### 7. Shallow Equality Bypass

Manual sync operations bypass shallow equality checks. When developers use direct writes, determining what has changed is their responsibility.

### 1. Optimistic Updates with Server Response

```typescript
const todosCollection = createCollection(queryCollectionOptions({
  queryKey: ['todos'],
  queryFn: fetchTodos,
  queryClient,
  getKey: (todo) => todo.id,
  
  onInsert: async ({ transaction, collection }) => {
    const newTodo = transaction.mutations[0].modified
    
    // Send to server
    const serverTodo = await api.createTodo(newTodo)
    
    // Server added fields like createdAt, updatedAt
    // Manually sync the server response (validated as update since it exists)
    collection.utils.syncUpdate(serverTodo)
    
    // Skip automatic refetch
    return { refetch: false }
  }
}))
```

### 2. WebSocket Integration with Error Handling

```typescript
// Initial collection setup
const messagesCollection = createCollection(queryCollectionOptions({
  queryKey: ['messages', roomId],
  queryFn: () => api.getMessages(roomId),
  queryClient,
  getKey: (msg) => msg.id,
}))

// Ensure initial data is loaded before setting up real-time updates
await messagesCollection.preload()

// WebSocket real-time updates with error handling
websocket.on('message:new', (message) => {
  try {
    messagesCollection.utils.syncInsert(message)
  } catch (error) {
    if (error.message.includes('existing key')) {
      // Message already exists, update instead
      messagesCollection.utils.syncUpdate(message)
    } else {
      console.error('Failed to sync new message:', error)
    }
  }
})

websocket.on('message:updated', (message) => {
  try {
    messagesCollection.utils.syncUpdate(message)
  } catch (error) {
    if (error.message.includes('non-existent')) {
      // Message doesn't exist yet, insert instead
      messagesCollection.utils.syncInsert(message)
    } else {
      console.error('Failed to sync message update:', error)
    }
  }
})

websocket.on('message:deleted', (messageId) => {
  try {
    messagesCollection.utils.syncDelete(messageId)
  } catch (error) {
    // Ignore if message already doesn't exist
    if (!error.message.includes('non-existent')) {
      console.error('Failed to sync message deletion:', error)
    }
  }
})
```

### 3. Incremental Loading / Pagination

```typescript
const productsCollection = createCollection(queryCollectionOptions({
  queryKey: ['products', { page: 1 }],
  queryFn: ({ queryKey }) => {
    const [, { page }] = queryKey
    return api.getProducts({ page, limit: 20 })
  },
  queryClient,
  getKey: (product) => product.id,
}))

// Ensure initial page is loaded
await productsCollection.preload()

// Load more products without refetching existing ones
async function loadMoreProducts() {
  const nextPage = currentPage + 1
  const newProducts = await api.getProducts({ page: nextPage, limit: 20 })
  
  // Filter out any products that might already exist (e.g., due to race conditions)
  const existingIds = new Set(
    Array.from(productsCollection.values()).map(p => p.id)
  )
  const uniqueNewProducts = newProducts.filter(p => !existingIds.has(p.id))
  
  if (uniqueNewProducts.length > 0) {
    // Add new products to collection without refetch
    productsCollection.utils.syncInsert(uniqueNewProducts)
  }
  
  // Optionally update the query cache for the next page
  queryClient.setQueryData(['products', { page: nextPage }], newProducts)
}
```

### 4. Safe Batch Sync with Conflict Resolution

```typescript
// Handle complex sync scenarios with proper error handling
async function syncBatchWithConflictResolution(changes: RemoteChanges) {
  const { inserts, updates, deletes } = changes
  
  // Pre-validate and separate items
  const validInserts: Item[] = []
  const validUpdates: Item[] = []
  const conflictingUpserts: Item[] = []
  
  // Check inserts
  for (const item of inserts) {
    if (collection.has(getKey(item))) {
      conflictingUpserts.push(item)
    } else {
      validInserts.push(item)
    }
  }
  
  // Check updates
  for (const item of updates) {
    if (!collection.has(getKey(item))) {
      conflictingUpserts.push(item)
    } else {
      validUpdates.push(item)
    }
  }
  
  // Execute batch with validated operations
  collection.utils.syncBatch([
    { type: 'insert', items: validInserts },
    { type: 'update', items: validUpdates },
    { type: 'upsert', items: conflictingUpserts },
    { type: 'delete', keys: deletes }
  ])
}
```

## Benefits

1. **Performance**: Avoid unnecessary full dataset refetches
2. **Real-time Support**: Easy integration with WebSockets/SSE
3. **Flexibility**: Support complex sync patterns (pagination, partial updates)
4. **Developer Experience**: Intuitive API consistent with collection patterns
5. **Consistency**: Maintains sync between collection and query cache
6. **Type Safety**: Full TypeScript support with proper overloads
7. **Data Integrity**: Runtime validation prevents data corruption
8. **Clear Error Messages**: Helpful errors guide developers to correct usage

## Implementation Notes

1. **Query Cache Updates**: Manual sync operations update the query cache by default, maintaining consistency with normal sync behavior

2. **Collection State Requirement**: Manual sync methods throw if collection is not in `ready` state. Developers must call `await collection.preload()` first

3. **Runtime Validation**: 
   - Insert operations validate no duplicate keys exist
   - Update operations validate items exist before updating
   - Delete operations validate items exist before deleting
   - Batch operations validate no conflicting operations
   - All operations validate keys are non-null/undefined

4. **Schema Validation**: Manual sync operations validate against the schema if provided

5. **Error Handling**: All validation errors throw with clear, actionable error messages

6. **TypeScript Overloads**: Provides overloads for single item vs array operations

7. **Shallow Equality**: Manual sync bypasses shallow equality checks - developers determine what has changed

## Migration Path

This feature is purely additive - no breaking changes required. Existing collections will continue to work as before, with new manual sync methods available as utilities.

## Implementation Priority

Given that this addresses common real-world use cases (WebSockets, server-computed fields, pagination), this should be considered a high-priority enhancement for the query-db-collection package.

### PR DESCRIPTION
**Pull request by KyleAMathews**
## Description

This PR adds manual write methods to `@tanstack/query-db-collection` to support direct state updates from external sources like WebSockets, server-sent events, or mutation responses.

## Problem

Currently, @tanstack/query-db-collection has two limitations:
1. **Performance**: Automatic refetching after mutations is inefficient for large datasets
2. **Real-time updates**: No way to manually sync data received through WebSockets or other channels

## Solution

Added new `write*` methods that bypass the normal optimistic update flow and directly write to collection state:

- `writeInsert(data)` - Write insert directly to collection
- `writeUpdate(data)` - Write update directly to collection  
- `writeDelete(keys)` - Write delete directly to collection
- `writeUpsert(data)` - Write upsert directly to collection
- `writeBatch(operations)` - Write batch operations directly

The `write` prefix follows established patterns from Apollo Client (`writeQuery`, `writeFragment`) and clearly indicates these methods bypass the normal persistence flow.

## Example Usage

```typescript
// WebSocket integration
ws.on('todo:created', (todo) => {
  todosCollection.utils.writeInsert(todo)
})

ws.on('todo:updated', (todo) => {
  todosCollection.utils.writeUpdate(todo)
})

ws.on('todo:deleted', (id) => {
  todosCollection.utils.writeDelete(id)
})

// Batch updates
ws.on('todos:sync', (changes) => {
  todosCollection.utils.writeBatch(
    changes.map(change => ({
      type: change.op,
      data: change.data
    }))
  )
})
```

## Implementation Details

- Extracted write logic into a separate `manual-sync.ts` module for better organization
- Unified operation handling eliminates code duplication
- Proper TypeScript types throughout
- Comprehensive test coverage (98.5%)
- All write operations update the TanStack Query cache after committing

## Breaking Changes

None - this is a new feature that doesn't affect existing APIs.

## Testing

Added comprehensive test suite covering:
- All write operations (insert, update, delete, upsert, batch)
- Error handling and validation
- Query cache synchronization
- Edge cases and batch operations

Closes #294

### CODE REVIEW

**Comment by jmarbutt** [comment]
This looks really good to me, I am excited to see this in and put it through the paces a little more.

File: packages/query-db-collection/src/query.ts [hunk 1]
```diff
@@ -222,8 +232,32 @@ export type RefetchFn = () => Promise<void>
...
+  TItem extends object = Record<string, unknown>,
+  TKey extends string | number = string | number,
+  TInsertInput extends object = TItem,
+> extends UtilsRecord {
   refetch: RefetchFn
+  syncInsert: (data: TInsertInput | Array<TInsertInput>) => void
+  syncUpdate: (updates: Partial<TItem> | Array<Partial<TItem>>) => void
+  syncDelete: (keys: TKey | Array<TKey>) => void
+  syncUpsert: (data: Partial<TItem> | Array<Partial<TItem>>) => void
+  syncBatch: (
```
Comment (kevin-dp) [inline packages/query-db-collection/src/query.ts]: I would simplify this public interface here. Since each sync operation is tagged with its type, we can expose the following public interface:
```ts
sync(op: SyncOperation<TItem, TKey, TInsertInput>)
syncBatch(ops: Array<SyncOperation<TItem, TKey, TInsertInput>>)
```

Internally, we can keep our `syncInsert`, `syncUpdate`, `syncDelete`, and `syncUpsert` methods and `sync` simply dispatches to the right one based on the type of operation.

File: packages/query-db-collection/src/query.ts [hunk 2]
```diff
@@ -236,9 +270,13 @@ export function queryCollectionOptions<
...
   TError = unknown,
   TQueryKey extends QueryKey = QueryKey,
+  TKey extends string | number = string | number,
+  TInsertInput extends object = TItem,
 >(
   config: QueryCollectionConfig<TItem, TError, TQueryKey>
-): CollectionConfig<TItem> & { utils: QueryCollectionUtils } {
+): CollectionConfig<TItem> & {
+  utils: QueryCollectionUtils<TItem, TKey, TInsertInput>
+} {
```
Comment (kevin-dp) [inline packages/query-db-collection/src/query.ts:265]: This function is 594 LoC long, that's not acceptable. We should take care to break this one down into smaller helper functions.

File: packages/query-db-collection/src/query.ts [hunk 3]
```diff
@@ -408,6 +446,371 @@ export function queryCollectionOptions<
...
+  // Store references to sync functions for manual sync operations
+  let syncFunctions: {
+    begin: () => void
+    write: (message: Omit<ChangeMessage<TItem>, `key`>) => void
+    commit: () => void
+    collection: any
+  } | null = null
+
+  // Enhanced internalSync that captures sync functions for manual use
+  const enhancedInternalSync: SyncConfig<TItem>[`sync`] = (params) => {
```
Comment (kevin-dp) [inline packages/query-db-collection/src/query.ts:447]: No need for this function, it adds unnecessary indirections. `internalSync` is only being called here on L465 and `enhancedInternalSync` is just being returned at the end of this function. So we can remove `enhancedInternalSync` and instead modify `internalSync` such that it assigns to `syncFunctions`.

File: packages/query-db-collection/src/query.ts [hunk 4]
```diff
@@ -408,6 +446,371 @@ export function queryCollectionOptions<
     })
   }
 
+  // Store references to sync functions for manual sync operations
+  let syncFunctions: {
```
Comment (kevin-dp) [inline packages/query-db-collection/src/query.ts]: Why do we need to use this pattern of receiving the sync functions later (when someone calls `sync`) and then assign it here and have to complicate the code with `if (!syncFunction) throw ...` checks everywhere?

Can't we just require the caller to pass the sync functions directly in the `config` argument of this `queryCollectionOptions` function? That would simplify a lot.

File: packages/query-db-collection/src/query.ts [hunk 5]
```diff
@@ -408,6 +446,371 @@ export function queryCollectionOptions<
...
+
+        case `update`: {
+          try {
+            key = getKey(operation.data as TItem) as TKey
+          } catch (error) {
+            throw new InvalidSyncOperationError(String(error))
+          }
+
+          // Check for duplicate keys within batch
+          if (seenKeys.has(key)) {
```
Comment (kevin-dp) [inline packages/query-db-collection/src/query.ts]: Why is this not allowed? Imagine i receive operations via a websocket, i buffer them, and then periodically call `syncBatch`. It's perfectly valid to receive an insert and then a bit later a delete which would then be processed in one batch. Why is this not allowed?

I will go a step further and define the following correctness claim for `syncBatch`:
> In the absence of concurrent operations, the result of sync batch should be equivalent to the result of syncing the operations from the batch sequentially.

But the current implementation violates this principle because this is allowed:
```ts
syncInsert({ id: `1`, a: 5 })
syncUpdate({ id: `1`, a: 6 })
```
But this isn't:
```ts
// This throws
syncBatch([
  { type: `insert`, data: { id: `1`, a: 5 } }
  { type: `update`, data: { id: `1`, a: 6 } }
])
```

File: packages/query-db-collection/src/query.ts [hunk 6]
```diff
@@ -408,6 +446,371 @@ export function queryCollectionOptions<
...
+
+    commit()
+
+    // Update query cache to reflect the new state
+    const currentData = collection.toArray as Array<TItem>
+    queryClient.setQueryData(queryKey, currentData as any)
+  }
+
   // Create wrapper handlers for direct persistence operations that handle refetching
   const wrappedOnInsert = onInsert
```
Comment (kevin-dp) [inline packages/query-db-collection/src/query.ts:471]: `wrappedOnInsert`, `wrappedOnUpdate`, and `wrappedOnDelete` are almost identical, lots of code duplication here. We should refactor this by introducing a `wrap` function that takes onInsert/onUpdate/onDelete and wraps it accordingly.

**Comment by KyleAMathews** [comment]
@kevin-dp give it another look please!

**Review by kevin-dp — APPROVED** [review]
